refine reduction from `HOL` to `HOL-Nat`

BilelGho opened the pull request:
`HOL-nat` is a specification over HOL functions. a  `HOL` function or definition is `HOL-nat` if the argument(s) type as well as the return type are exclusively `nat` and all the call-back functions are HOL.

A refined function takes the same arguments as the original function,encoded as natural numbers, and returns the same the result, also encoded as a natural number.

In this branch a refinement of the already developed the reduction of a IMP- program (a polynomially bounded verificator for a NP problem) to a SAT formula is implemented. The equivalence between the initial function and the refined one is verified.

BilelGho committed:
Merge pull request #1 from wimmers/master

Imp minus (#23)

BilelGho committed:
Merge branch 'wimmers:master' into master

BilelGho committed:
old ROOTS

BilelGho committed:
Merge branch 'master' of github.com:BilelGho/poly-reductions

BilelGho committed:
non-harmful changes to ease the refinement to HOL-nat

BilelGho committed:
refinement of the reduction to HOL_nat

BilelGho committed:
refinement of the reduction to HOL_nat

BilelGho committed:
refinement of the reduction to HOL_nat

BilelGho committed:
Merge branch 'wimmers:master' into master

BilelGho committed:
Merge pull request #2 from wimmers/cook-levin-hol-statement

Added final HOL version of the Cook-Levin thm

mabdula commented:
Does this have a nat version of the theorem main_hol_lemma? I couldn't find that

notiho committed:
slightly cleaned up proofs

BilelGho committed:
push everything to work on remote

BilelGho committed:
ROOT file

BilelGho committed:
forgot to add file

Commit:
reduction no longer existentially quantified + refined called functions to HOL-computable

Commit:
forgot to add file

Commit:
Merge branch 'master' of github.com:BilelGho/poly-reductions

notiho committed:
implemented and verified multiplication in IMP-

Commit:
refinement to HOL_nat + main_lemma_hol_nat

Commit:
fixed ;)

Commit:
Merge branch 'wimmers-master'

Commit:
transformed higher order functions into independent instances

Commit:
started refinment to HOL-tail

Commit:
refinement to tail-rec, started general pattern for IMP- to IMP--

BilelGho committed:
IMP- to IMP-- tail recursive

Commit:
prove correctness for nat, imp- to imp--, first try

Commit:
IMP- to IMP -- refined to tail recursive

BilelGho committed:
all the functions that need a general pattern for tail recursion

Commit:
more refinement to tail-rec , max constant remaining

BilelGho committed:
max constant refined to tail-rec

Commit:
finished refinement to tail-rec

notiho committed:
updated multiplication

notiho committed:
Merge branch 'wimmers:master' into multiplication

notiho committed:
Merge pull request #1 from BilelGho/master

tail recursive

notiho committed:
verified implementation of encoding and decoding pairs in IMP-

notiho committed:
implemented and verified a few functions from IMP- Max Constant

Commit:
termination proof: trying with size function

BilelGho committed:
temrination order for max_constant, and all_variables (stack versions ), yet nat version of those aare missing termination proofs

BilelGho committed:
temrination order for stack simulation implementations on HOL level, HOL-nat not yet

notiho committed:
implemented and verified add_res

BilelGho committed:
show termination for nat, only for valid input and replace correctness proofs with psimps

BilelGho committed:
rewrite prod_encode, fst_nat , snd_nat in polynomial time  code (while staying tail recursive), with functional correctness proof

notiho committed:
started implementing max_constant

notiho committed:
Merge pull request #2 from BilelGho/master

m

notiho committed:
finished implementing and verifiying max_constant

notiho committed:
implemented another function

Commit:
Created empty theory IMP_Minus_Common_Funs_Nat

Commit:
imported IMP_Minus_Nat_Bijection

Commit:
stopped working on append

Commit:
defined elemof_IMP_Minus_iteration

Commit:
defined elemof_IMP_Minus_iteration_time

Commit:
stated and gave very messy proof to elemof_IMP_Minus_iteration_correct

Commit:
Defined elemof_IMP_Minus_loop

Commit:
Defined elemof_IMP_Minus_loop_time

Commit:
Stated and gave horrible proof to elemof_IMP_Minus_loop_correct

lakiryt committed:
removed unnecessary simp add

lakiryt committed:
defined elemof_IMP_Minus

lakiryt committed:
defined elemof_IMP_Minus_time

lakiryt committed:
Stated and gave awfully verbose proof elemof_IMP_Minus_correct

lakiryt committed:
marginally shortened elemof_IMP_Minus_correct proof

lakiryt committed:
marginally shortened elemof_IMP_Minus_loop_correct

lakiryt committed:
shortened True-case in elemof_IMP_Minus_iteration_correct

lakiryt committed:
shortened proof for elemof_IMP_Minus_iteration_correct

lakiryt committed:
simplified time definition

lakiryt committed:
shortened elemof_IMP_Minus_correct proof

lakiryt committed:
removed lemmata from elemof_IMP_Minus_loop_correct

lakiryt committed:
stylistics

lakiryt committed:
Defined append_tail_IMP_Minus

lakiryt committed:
defined append_tail_IMP_Minus_time

lakiryt committed:
named revapp lemma

lakiryt committed:
stated and (verbosely) proved append_tail_IMP_Minus_correct

lakiryt committed:
shortened proof of append_tail_IMP_Minus_correct

lakiryt committed:
removed unused lemma

lakiryt committed:
simplified append_tail_IMP_Minus_time

lakiryt committed:
stopped working on remdups_tail due to run out of registers

lakiryt committed:
defined list_from_acc_IMP_Minus_iteration

lakiryt committed:
defined list_from_acc_IMP_Minus_iteration_time

lakiryt committed:
stated and proved list_from_acc_IMP_Minus_iteration_correct

lakiryt committed:
defined list_from_acc_IMP_Minus_loop_time

lakiryt committed:
stated and proved list_from_acc_loop_correct

lakiryt committed:
shortened proof for list_from_acc_correct

lakiryt committed:
defined concat_acc_IMP_Minus_iteration with additional registers

lakiryt committed:
defined concat_acc_IMP_Minus_iteration_time

lakiryt committed:
stated and proved concat_acc_IMP_Minus_iteration_correct

lakiryt committed:
defined concat_acc_IMP_Minus_loop_time

lakiryt committed:
stated and proved concat_acc_IMP_Minus_correct

notiho committed:
added canonical state transformers

lakiryt committed:
Merge remote-tracking branch 'upstream/multiplication' into utilfun

notiho committed:
started changing IMP- implementations to state_transformer

lakiryt committed:
Merge remote-tracking branch 'upstream/multiplication' into utilfun

lakiryt committed:
left proof with "sorry"s

lakiryt committed:
removed unnecessary parts

lakiryt committed:
proved c1

lakiryt committed:
proved c2

lakiryt committed:
marginally shortened c2 proof

lakiryt committed:
corrected overridden local variable name and corresponding proof steps

lakiryt committed:
re-added formerly working proof before sorry

lakiryt committed:
factored out duplicate auxiliary proof

lakiryt committed:
improved readability of the cons_list_IMP_Minus_state_transformer and reduced proof length (esp. aux)

lakiryt committed:
further shortened proof, gave meaningful name to aux->cons_list_state_arv

lakiryt committed:
corrected bracketing, reviving Nil-case proof (+marginally shortened)

lakiryt committed:
removed commented out old/experimental state_transformers for cons_list

BilelGho committed:
Merge pull request #4 from lakiryt/utilfun

Refinement to IMP- using namespaces

BilelGho committed:
working on first function in binary arithmetic theory

stuck on final state proof.

BilelGho committed:
fixing state transformers

composition was applied in the wrong direction + not all side effects were zeroed-out

BilelGho committed:
test

mabdula committed:
Added experimental proofs to deal with IH, if and while

mabdula committed:
New approach instaed of state transformers

mabdula committed:
Finished prod_encode

mabdula committed:
Mul is standard now

mabdula committed:
prod decode

mabdula committed:
hd and tl

mabdula committed:
IMP- primitives in a new file

mabdula committed:
Call by prefixes in IMP-

mabdula committed:
Moved IMP- primitives to new file; Refined some more primitives; New session for Cook Levin Nat

mabdula committed:
Added some comment

mabdula committed:
N-th bit + slightly better infrastructure
